File: Python3/0865_smallest_subtree_with_all_the_deepest_nodes.py
# https://leetcode.cn/problems/smallest-subtree-with-all-the-deepest-nodes/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def subtreeWithAllDeepest(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        res = None
        depth_max = -1

        def dfs(node: Optional[TreeNode], depth: int) -> int:
            nonlocal res, depth_max
            if node is None:
                depth_max = max(depth_max, depth)
                return depth
            
            left_depth_max = dfs(node.left, depth+1)
            right_depth_max = dfs(node.right, depth+1)

            if left_depth_max == right_depth_max and left_depth_max == depth_max:
                res = node
            
            return max(left_depth_max, right_depth_max)

        dfs(root, 0)
        return res

File: Cpp/0865_smallest_subtree_with_all_the_deepest_nodes.cpp
// https://leetcode.cn/problems/smallest-subtree-with-all-the-deepest-nodes/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

#include <algorithm>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    TreeNode* subtreeWithAllDeepest(TreeNode* root) {
        TreeNode *res = nullptr;
        int depth_max = -1;

        auto dfs = [&](auto &&self, TreeNode *node, int depth) -> int {
            if (node == nullptr) {
                depth_max = max(depth_max, depth);
                return depth;
            }

            int left_depth_max = self(self, node->left, depth+1);
            int right_depth_max = self(self, node->right, depth+1);
            if (left_depth_max == right_depth_max && left_depth_max == depth_max) {
                res = node;
            }
            return max(left_depth_max, right_depth_max);
        };

        dfs(dfs, root, 0);
        return res;
    }
};